### PR TITLE
add resnet20 model, clean up cifar logs

### PR DESCRIPTION
**Pull request by brettkoonce**

### CODE REVIEW

File: CIFAR/main.swift [hunk 1]
```diff
@@ -27,15 +29,25 @@ for epoch in 1...10 {
...
+        let correctPredictions = logits.argmax(squeezingAxis: 1) .== labels
+        correctGuessCount = correctGuessCount + Int(Tensor<Int32>(correctPredictions).sum().scalarized())
+        totalGuessCount = Int32(totalGuessCount) + Int32(batchSize)
     }
-    print("  average loss: \(testLossSum / Float(testBatchCount))")
+
+    let accuracy = Float(correctGuessCount) / Float(totalGuessCount)
+    print("""
+          [Epoch \(epoch)] \
+          Accuracy: \(correctGuessCount)/\(totalGuessCount) (\(accuracy)) loss: \(testLossSum / Float(testBatchCount))
```
Comment (rxwei) [inline CIFAR/main.swift]: Fit lines within 100 columns.

File: CIFAR/main.swift [hunk 2]
```diff
@@ -27,15 +29,25 @@ for epoch in 1...10 {
...
+    var correctGuessCount:Int = 0
+    var totalGuessCount:Int32 = 0
     for batch in testBatches {
         let (labels, images) = (batch.first, batch.second)
         let logits = model.inferring(from: images)
         testLossSum += softmaxCrossEntropy(logits: logits, labels: labels).scalarized()
         testBatchCount += 1
+
+        let correctPredictions = logits.argmax(squeezingAxis: 1) .== labels
+        correctGuessCount = correctGuessCount + Int(Tensor<Int32>(correctPredictions).sum().scalarized())
```
Comment (rxwei) [inline CIFAR/main.swift]: Fit within 100 columns.

File: CIFAR/resnet.swift [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    var basicBlock1: BasicBlock
+    var basicBlock2: BasicBlock
+    var basicBlock3: BasicBlock
+
+    var averagePool: AvgPool2D<Float>
+    var flatten = Flatten<Float>()
+    var classifier: Dense<Float>
+
+    public init() {
+        inputLayer = Conv2DBatchNorm(filterShape: (3, 3, 3, 16), padding: .same)
```
Comment (rxwei) [inline CIFAR/resnet.swift]: Since the initialization does not depend on any initializer arguments, you can move these initialization expressions inline like how you initialized `flatten`, and delete `init()`.

File: CIFAR/resnet.swift [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return tmp
+    }
+}
+
+struct BasicBlock: Layer {
+    var layer1: Conv2DBatchNorm
+    var layer2: Conv2DBatchNorm
+    var shortcut: Conv2DBatchNorm
+
+    public init(
```
Comment (rxwei) [inline CIFAR/resnet.swift]: The type is internal. Remove `public`.
```suggestion
   init(
```

File: CIFAR/resnet.swift [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        filterShape: (Int, Int, Int, Int),
+        strides: (Int, Int) = (1, 1),
+        padding: Padding = .valid
+    ) {
+        self.conv = Conv2D(filterShape: filterShape, strides: strides, padding: padding)
+        self.norm = BatchNorm(featureCount: filterShape.3)
+    }
+
+    @differentiable
+    public func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
```
Comment (rxwei) [inline CIFAR/resnet.swift]:
```suggestion
    func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
```

File: CIFAR/resnet.swift [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+/// Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
+/// https://arxiv.org/abs/1512.03385
+/// using shortcut layer to connect BasicBlock layers (aka Option (B))
+/// see https://github.com/akamaster/pytorch_resnet_cifar10 for explanation
+
+struct Conv2DBatchNorm: Layer {
+    var conv: Conv2D<Float>
+    var norm: BatchNorm<Float>
+
+    public init(
```
Comment (rxwei) [inline CIFAR/resnet.swift]:
```suggestion
    init(
```

File: CIFAR/resnet.swift [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    @differentiable
+    func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
+        var tmp = relu(layer1.applied(to: input, in: context))
+        tmp = relu(layer2.applied(to: tmp, in: context))
+        return relu(tmp + shortcut.applied(to: input, in: context))
+    }
+}
+
+public struct ResNet20: Layer {
```
Comment (rxwei) [inline CIFAR/resnet.swift]: No need to be public because it's only used within the current module, not exposed as an API.
```suggestion
struct ResNet20: Layer {
```

File: CIFAR/resnet.swift [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        basicBlock1 = BasicBlock(featureCounts:(16, 16, 16, 16), strides: (1,1))
+        basicBlock2 = BasicBlock(featureCounts:(16, 32, 32, 32))
+        basicBlock3 = BasicBlock(featureCounts:(32, 64, 64, 64))
+
+        averagePool = AvgPool2D(poolSize: (8, 8), strides: (8, 8))
+        classifier = Dense(inputSize: 64, outputSize: 10, activation: softmax)
+    }
+
+    @differentiable
+    public func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
```
Comment (rxwei) [inline CIFAR/resnet.swift]:
```suggestion
    func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
```

File: CIFAR/main.swift [hunk 9]
```diff
@@ -27,15 +29,27 @@ for epoch in 1...10 {
...
+    var totalGuessCount:Int32 = 0
     for batch in testBatches {
         let (labels, images) = (batch.first, batch.second)
         let logits = model.inferring(from: images)
         testLossSum += softmaxCrossEntropy(logits: logits, labels: labels).scalarized()
         testBatchCount += 1
+
+        let correctPredictions = logits.argmax(squeezingAxis: 1) .== labels
+        correctGuessCount = correctGuessCount + Int(Tensor<Int32>(correctPredictions)
+          .sum().scalarized())
```
Comment (rxwei) [inline CIFAR/main.swift]: * Indent by 4 from the last line.
* It would be better to break the line after `+` and put the entire `Int(...)` expression at the bottom.

File: CIFAR/main.swift [hunk 10]
```diff
@@ -27,15 +29,27 @@ for epoch in 1...10 {
         trainingBatchCount += 1
         optimizer.update(&model.allDifferentiableVariables, along: gradients)
     }
-    print("  average loss: \(trainingLossSum / Float(trainingBatchCount))")
-    print("Epoch \(epoch), evaluating on test set...")
+
     var testLossSum: Float = 0
     var testBatchCount = 0
+    var correctGuessCount:Int = 0
```
Comment (rxwei) [inline CIFAR/main.swift]: Integer literals default to `Int`.

```suggestion
    var correctGuessCount = 0
```

File: CIFAR/main.swift [hunk 11]
```diff
@@ -27,15 +29,27 @@ for epoch in 1...10 {
...
+    var totalGuessCount:Int32 = 0
```
Comment (rxwei) [inline CIFAR/main.swift]: Add a space after colons.

```suggestion
    var totalGuessCount: Int32 = 0
```

File: CIFAR/resnet.swift [hunk 12]
```diff
@@ -0,0 +1,90 @@
+import TensorFlow
+
+/// Original Paper:
```
Comment (rxwei) [inline CIFAR/resnet.swift]: Switch to double-slashed comments because it is not a doc comment block for type `Conv2DBatchNorm`.

File: CIFAR/main.swift [hunk 13]
```diff
@@ -27,15 +29,27 @@ for epoch in 1...10 {
...
     for batch in testBatches {
         let (labels, images) = (batch.first, batch.second)
         let logits = model.inferring(from: images)
         testLossSum += softmaxCrossEntropy(logits: logits, labels: labels).scalarized()
         testBatchCount += 1
+
+        let correctPredictions = logits.argmax(squeezingAxis: 1) .== labels
+        correctGuessCount = correctGuessCount + Int(Tensor<Int32>(correctPredictions)
+          .sum().scalarized())
+        totalGuessCount = Int32(totalGuessCount) + Int32(batchSize)
```
Comment (rxwei) [inline CIFAR/main.swift]: Aren't these already `Int32`?
```suggestion
        totalGuessCount = totalGuessCount + batchSize
```

**Comment by brettkoonce** [comment]
@rxwei kk trying to keep up let me do some testing before you merge this!

File: CIFAR/resnet.swift (hunk 3 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: done

File: CIFAR/main.swift (hunk 9 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: kk

File: CIFAR/main.swift (hunk 13 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: yes, removed

File: CIFAR/resnet.swift (hunk 12 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed leading slash

File: CIFAR/main.swift (hunk 11 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: done

File: CIFAR/main.swift (hunk 10 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: removed

File: CIFAR/resnet.swift (hunk 8 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed all public calls

File: CIFAR/resnet.swift (hunk 7 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed all public calls

File: CIFAR/resnet.swift (hunk 6 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed all public calls

File: CIFAR/resnet.swift (hunk 5 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed all public calls

File: CIFAR/resnet.swift (hunk 4 above)
Comment (brettkoonce) [reply CIFAR/resnet.swift]: removed all public calls

File: CIFAR/main.swift (hunk 2 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: added newline

Comment (brettkoonce) [reply CIFAR/main.swift]: reindented per feedback below

File: CIFAR/main.swift (hunk 1 above)
Comment (brettkoonce) [reply CIFAR/main.swift]: added newline before Loss

**Comment by brettkoonce** [comment]
@rxwei looking better, let me know what you think!

**Comment by brettkoonce** [comment]
🚀

**Comment by johndpope** [comment]
Is there a swift lint schema that could automate the indentation ? would save some time conforming to standards.

**Comment by johndpope** [comment]
https://github.com/realm/SwiftLint Probably just need to craft rules to match google style guide. I wonder if there’s already being used.

**Comment by rxwei** [comment]
`swift-format`
https://github.com/apple/swift/blob/master/docs/SwiftFormat.md
